File: financials/services/accounts-service.js
import fetch from "node-fetch";
import pool from "../db/pgPool.js";

class AccountsService {

    constructor(url) {
        this.url = url
    }

    async pullData() {
        const res = await fetch(this.url,{
            method: 'GET',
            headers: {
                'X-AgreementGrantToken': 'demo',
                'X-AppSecretToken': 'demo'
            }
        })
        return await res.json();
    }


    async pullEntries(accountNumber,url){
        const resYears = await this.pullAccountYears(accountNumber)
        const years = resYears.collection.map(({year})=>year)
        years.forEach(async(year)=>{
            const res = await fetch('https://restapi.e-conomic.com/accounts/'+accountNumber+'/accounting-years/'+year+'/entries',{
            method: 'GET',
            headers: {
                'X-AgreementGrantToken': 'demo',
                'X-AppSecretToken': 'demo'
            }
            })
            const entries = (await res.json()).collection
            entries.forEach(async(entry)=>{
                const {
                    account
                    ,amount
                    ,amountInBaseCurrency
                    ,currency
                    ,date
                    ,dueDate
                    ,departmentalDistribution
                    ,entryNumber
                    ,text
                    ,entryType
                    ,vatAccount
                    ,voucherNumber
                    ,quantity1
                    ,quantity2
                    ,bookedInvoice
                    ,invoiceNumber  
                } = entry;
               
                const insertScript = `INSERT INTO entries(
                    account,
                    amount,
                    amountInBaseCurrency,
                    currency,
                    date,
                    dueDate,
                    departmentalDistribution,
                    entryNumber,
                    text,
                    entryType,
                    vatAccount,
                    voucherNumber,
                    quantity1,
                    quantity2,
                    bookedInvoice,
                    invoiceNumber
                ) VALUES (
                    ${account.accountNumber}
                        ,${amount}
                        ,${amountInBaseCurrency}
                        ,'${currency}'
                        ,${date?"'"+date+"'":'NULL'}
                        ,${dueDate?"'"+dueDate+"'":'NULL'}
                        ,${departmentalDistribution?departmentalDistribution.departmentalDistributionNumber:'NULL'}
                        ,${entryNumber}
                        ,'${text}'
                        ,'${entryType}'
                        ,'${vatAccount?vatAccount.vatCode:'NULL'}'
                        ,${voucherNumber}
                        ,${quantity1?quantity1:'NULL'}
                        ,${quantity2?quantity2:'NULL'}
                        ,${bookedInvoice?bookedInvoice.bookedInvoiceNumber:'NULL'}
                        ,${invoiceNumber?invoiceNumber:'NULL'}
                );`
                return await pool.query(insertScript,[])

            })
        })
        
    }

    async pullAccountYears(accountNumber){
        const res = await fetch('https://restapi.e-conomic.com/accounts/'+accountNumber+'/accounting-years',{
            method: 'GET',
            headers: {
                'X-AgreementGrantToken': 'demo',
                'X-AppSecretToken': 'demo'
            }
        })
        return await res.json();
    }

    async uploadData(data) {
        
        data.collection.forEach(async(account) => {
            const {
                accountNumber,
                accountType,
                balance,
                blockDirectEntries,
                debitCredit,
                name,
            } = account;
            
            const insertScript = `INSERT INTO accounts(
                accountNumber,
                accountType,
                balance,
                blockDirectEntries,
                debitCredit,
                name) VALUES (
                    ${accountNumber},
                    '${accountType}',
                    ${balance},
                    '${blockDirectEntries}',
                    '${debitCredit}',
                    '${name}'
                )`;
            const response = await pool.query(insertScript,[])
            this.pullEntries(accountNumber,'').catch(err=>{console.log(err)})

            return response

        });
        
    }

    async insertData() {
        const savePullData = await this.pullData()
        const saveUploadData = await this.uploadData(savePullData)
        //return saveUploadData
    }

}

export default AccountsService;

File: financials/services/accounts_service.py
import json
import requests

from sqlalchemy.orm import Session

from models.account import Account

class AccountsService:
    """
    Class handling the pulling of data from
    e-conomic and saving it to db
    """
    def __init__(self,url,engine):
        self.url = url
        self.engine = engine

    def pull_data(self):
        """
        fetches data
        """
        res = requests.get(
            self.url,
            headers={
                'X-AgreementGrantToken': 'demo',
                'X-AppSecretToken': 'demo'
            }
        )
        return json.loads(res.text)

    def upload_data(self,data):
        """
        Uploads data in case of conflict of PK it updates
        """
        with Session(self.engine) as session:
            for account in data['collection']:
                query = session.query(Account).where(
                    Account.account_number == account['accountNumber']
                )
                existing_account = query.first()
                if existing_account:
                    existing_account.account_type = account['accountType']
                    existing_account.balance = account['balance']
                    existing_account.block_direct_entries = account['blockDirectEntries']
                    existing_account.debit_credit = account['debitCredit']
                    existing_account.name = account['name']
                    session.add(existing_account)
                else:
                    new_account = Account(
                        account_number = account['accountNumber'],
                        account_type = account['accountType'],
                        balance = account['balance'],
                        block_direct_entries= account['blockDirectEntries'],
                        debit_credit = account['debitCredit'],
                        name = account['name'],
                    )
                    session.add(new_account)
            session.commit()

    def insert_data(self):
        """
        calls sub routines for fetching and storing
        """
        data = self.pull_data()
        self.upload_data(data)
